04_Scheduling/03_Tasks_And_Threads.md

```diff
@@ -0,0 +1,132 @@
+    function(arg);
+    _thread_exit();
+}
+```
+
+What should the exit function do? Again this depends on the design choice, and as usual there are multiple paths, it depends if we want to exit the thread as soon as it calls the the exit function, or let the scheduler do that, just updating its status to the one correspinding to a terminated task (let's call it DEAD status). 
+
+* If the choice is to delete the thread as soon as it exit terminate the function what will happen then is that the thread is placed into a DEAD state, then the function will take care of removing the task from the scheduler queue, freeing all resources (stack, execution frame, page tables, etc.) that are allocated to it, and after that it will free the memory allocated to the task itself. Remember that when a task call the exit function is still the one being executed
+* In the other scenario, what the exit function does is basically just updating the status of the task to DEAD. And not much more. Then next time the scheduler will be called, it will pick the next task, execute it, and so on, after sometime it will pick up again the terminated task, it will see the status as DEAD, so it will start the same process explained above, free the associated resources to the thread, remove it from the queue, and free the thread item. After that the scheduler will pick the next task, if the status will be not DEAD it will prepare it for the execution
+So when a function is called we have: 
+
+* The first 6 parameters stored into the rdi, rsi, rdx, rcx, r8 to r15 registers
+* The rest of the parameters are pushed on the stack
+* The return address is pushed on the stack after the parameters.
+* rax (and eventually rbx) are used for the return value
+
+The third item in the bullet list is the one we are interested to. When a function is called in asm we have something like: 
+
+```asm 
+mov rdi, 5 // First parameter
+call _function
+; do something
+```
+
+Now when a function is  called, the cpu put the next instruction on the stack. And here is where we will put the end function.
+
+* Then the thread_exit function should take also care of updating the thread status to DEAD (and by extension if it was the last thread on a task it should update that too)
+* The schedule function when pick a task that has the status set to DEAD knows that it doesn´t have to execute it, and will call the routine to free the thread resources. 
+* If the thread is the last in a task, the task should be ready to be deleted too. So in this case we need to change the status to the task too. 
+ 
+## Switching tasks/thread
+
+So who is responsible of switching tasks? The scheduler, that every time is called it checks if the current task needs to be replaced with a new one, and eventually pick the "best" next task candidate. How to decide who is the next depends on the scheduling algorithm implemented and the design choice of the operating system (there are many scheduling algorithm). 
```

Comment:
This is a repeat of parts of the scheduling chapter. I'd also say that thread selection should be part of the scheduler, not tasks and threads.